lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -452,66 +503,141 @@ def name(self) -> str:
         return "opensearch-knn"
 
 
+class OpenSearchPluginBackupDataProvider(OpenSearchPluginDataProvider):
+    """Responsible to decide which data to use for the backup plugin.
+
+    Backups should check different relations depending on their role in the cluster:
+    * main orchestrator
+    * failover orchestrator
+    * other
+    """
+
+    def __init__(self, charm):
+        """Creates the OpenSearchPluginBackupDataProvider object."""
+        super().__init__(charm)
+        self._relation = None
+        if not self._charm.opensearch_peer_cm.deployment_desc():
+            # Temporary condition: we are waiting for CM to show up and define which type
+            # of cluster are we. Once we have that defined, then we will process.
+            raise OpenSearchPluginMissingConfigError("Missing deployment description in peer CM")
+
+        self.is_main_orchestrator = (
+            self._charm.opensearch_peer_cm.deployment_desc().typ
+            == DeploymentType.MAIN_ORCHESTRATOR
+        )
+
+    def get_relation(self) -> Any:
+        """Updates the relation object if needed."""
+        self._relation = self._charm.model.get_relation(S3_RELATION)
+        if not self.is_main_orchestrator:
+            self._relation = self._charm.model.get_relation(PeerRelationName)
+        return self._relation
+
+    def get_data(self) -> Dict[str, Any]:
+        """Returns the data from the relation databag.
+
+        Exceptions:
+            ValueError: if the data is not valid
+        """
+        if not self.get_relation():
+            return {}
+        result = dict(self.get_relation().data[self._relation.app]) or {}
+        if not self.is_main_orchestrator:
+            # Peer relations exchange secrets via peer-cluster secret
+            result |= self._charm.secrets.get_object(Scope.APP, S3_CREDENTIALS) or {}
+        return result
+
+
 class OpenSearchBackupPlugin(OpenSearchPlugin):
     """Manage backup configurations.
 
     This class must load the opensearch plugin: repository-s3 and configure it.
+
+    The plugin is responsible for managing the backup configuration, which includes relation
+    databag or only the secrets' content, as backup changes behavior depending on the juju app
+    role within the cluster.
     """
 
+    MODEL = S3RelData
+    MANDATORY_CONFS = [
+        "bucket",
+        "endpoint",
+        "region",
+        "base_path",
+        "protocol",
+        "credentials",
+    ]
+    DATA_PROVIDER = OpenSearchPluginBackupDataProvider
```

Comment:
why a class variable?

Reply:
They are very tightly coupled. The idea is to have each subclass defining its own provider type. To simplify it, I am stating a class and then the upstream classes creating the plugin do not need to worry to create 2x objects instead of one.